Add blog article about minReadySeconds for StatefulSets and maxSurge for DaemonSets (#36763)
kubernetes/website: content/en/blog/_posts/2022-07-27-sig-apps-GA-1.25/index.md

---
layout: blog
title: "sig-apps features graduating to stable in 1.25"
date: 2022-07-27
slug: "sig-apps features graduating to stable in 1.25"
---

**Authors:** Ravi Gudimetla (Apple), Filip Křepinský (Red Hat), Maciej Szulik (Red Hat)

This blog describes the two features namely `minReadySeconds` for StatefulSets and `maxSurge` for DaemonSets that sig-apps is happy to graduate to stable in 1.25.

Specifying `minReadySeconds` slows down a rollout of a StatefulSet when using a `RollingUpdate` `updateStrategy` by waiting for each pod for a desired time. 
This time can be used for initializing the pod (e.g. warming up the cache) or as a delay before acknowledging the pod.

`MaxSurge` allows a DaemonSet workload to run multiple instances of the same pod on a node during a rollout when using a `RollingUpdate` `updateStrategy`.
This helps to minimize the downtime of the DaemonSet to other consumers.

> These features are already available in a Deployment and other workloads. This graduation helps to align this functionality across the workloads.

Author: >  I'd start with a simple use-case describing the necessity for both of these features and also add d add that this effort is to align higher-level workloads controllers between each other, since Deployments already support both of these features. Only then go into details how this works and how is implemented.

@soltysh would you prefer to have also YAML's of the workloads here? I think that would be too verbose...

Reviewer: Not needed, just explanation what it is for. YAMLs are in the docs, and we can provide links to them, for those interested more in details.

> These both features are already available in a Deployment (and other workloads). This graduation helps to align this functionality across the workloads.

Author: > I'd start with a simple use-case describing the necessity for both of these features and also add d add that this effort is to align higher-level workloads controllers between each other, since Deployments already support both of these features. Only then go into details how this works and how is implemented.

@soltysh would you prefer to have also YAML's of the workloads here? I think that would be too verbose...

## How does it work?

### MinReadySeconds for StatefulSets

The StatefulSet controller watches for the StatefulSet pods and counts how long a particular pod has been in the `Running` state, if this value is greater than or equal to the time specified in `.spec.minReadySeconds` field of the StatefulSet, the StatefulSet controller updates the `AvailableReplicas` field in the StatefulSet's status.

### MaxSurge for DaemonSets

The DaemonSet controller creates the additional pods (above the desired number resulting from DaemonSet spec) based on the value given in `.spec.strategy.rollingUpdate.maxSurge`. The additional pods would run on the same node where the old DaemonSet pod is running till the old pod gets killed. This value cannot be `0` when `MaxUnavailable` is 0. The default value is 0.

## How do I use it?

### MinReadySeconds for StatefulSets

Specify a value for `minReadySeconds` for any StatefulSet and check if pods are available or not by inspecting
`AvailableReplicas` field using:

`kubectl get statefulset/<name_of_the_statefulset> -o yaml`

Please note that the default value of `minReadySeconds` is 0

### MaxSurge for DaemonSets

> Specify a value for `.spec.updateStrategy.rollingUpdate.maxSurge` and set `.spec.updateStrategy.rollingUpdate.maxUnavailable` to `0`. 

Author: `maxUnavailable` cannot be empty since it will get defaulted to `1` which will in turn fail validation (we cannot have both [maxSurge and maxUnavailable](https://github.com/kubernetes/kubernetes/blob/0f37b3120643580f632ca12d3e174e7ec447948c/pkg/apis/apps/validation/validation.go#L387). Should I mentioned this?

this was a bit surprise to me, because you have to remember to specify both.

---
layout: blog
title: "sig-apps features graduating to stable in 1.25"
date: 2022-07-27
slug: "sig-apps features graduating to stable in 1.25"
---

**Authors:** Ravi Gudimetla (Apple), Filip Křepinský (Red Hat), Maciej Szulik (Red Hat)

This blog describes the two features namely `minReadySeconds` for StatefulSets and `maxSurge` for DaemonSets that sig-apps is happy to graduate to stable in 1.25.

> Specifying `minReadySeconds` slows down a rollout of a StatefulSet when using a `RollingUpdate` `updateStrategy` by waiting for each pod for a desired time. 

Reviewer: ```suggestion
Specifying `minReadySeconds` slows down a rollout of a StatefulSet, when using a `RollingUpdate` value in `.spec.updateStrategy` field, by waiting for each pod for a desired time. 
```

This time can be used for initializing the pod (e.g. warming up the cache) or as a delay before acknowledging the pod.

> `maxSurge` allows a DaemonSet workload to run multiple instances of the same pod on a node during a rollout when using a `RollingUpdate` `updateStrategy`.

Reviewer: ```suggestion
`maxSurge` allows a DaemonSet workload to run multiple instances of the same pod on a node during a rollout when using a `RollingUpdate` value in `.spec.updateStrategy` field.
```

> This helps to minimize the downtime of the DaemonSet to other consumers.

Reviewer: ```suggestion
This helps to minimize the downtime of the DaemonSet for consumers.
```

These features are already available in a Deployment and other workloads. This graduation helps to align this functionality across the workloads.

> ## What problems does these features solve?

Reviewer 2: ```suggestion
## What problems do these features solve?
```

Author: fixed

### MinReadySeconds for StatefulSets

> `minReadySeconds` ensures that the StatefulSet workload is `Ready` for the given number of seconds before calling the

Reviewer: ```suggestion
`minReadySeconds` ensures that the StatefulSet workload is `Ready` for the given number of seconds before reporting the
```

> pod `Available`. The notion of being `Ready` and `Available` is quite important for workloads. For example, some workloads like Prometheus with multiple instances of Alertmanager should be considered `Available` only when the Alertmanager's state transfer is complete. `minReadySeconds` also helps when using loadbalancers with cloud providers. Since the pod should be `Ready` for the given number of seconds, it provides buffer time to prevent killing pods in rotation before new pods show up.

Reviewer: ```suggestion
pod `Available`. The notion of being `Ready` and `Available` is quite important for workloads. For example, some workloads, like Prometheus with multiple instances of Alertmanager, should be considered `Available` only when the Alertmanager's state transfer is complete. `minReadySeconds` also helps when using loadbalancers with cloud providers. Since the pod should be `Ready` for the given number of seconds, it provides buffer time to prevent killing pods in rotation before new pods show up.
```

### MaxSurge for DaemonSets

> Kubernetes system-level components like CNI, CSI are typically run as DaemonSets. These components can have impact on the availability of the workloads if those DaemonSets go down momentarily during the upgrades. The feature allows DaemonSet pods to surge, thereby ensuring zero-downtime for the DaemonSets.

Reviewer: ```suggestion
Kubernetes system-level components like CNI, CSI are typically run as DaemonSets. These components can have impact on the availability of the workloads if those DaemonSets go down momentarily during the upgrades. The feature allows DaemonSet pods to temporarily increase their number, thereby ensuring zero-downtime for the DaemonSets.
```

Please note that the usage of `HostPort` in conjunction with `MaxSurge` in DaemonSets is not allowed as DaemonSet pods are tied to a single node and two active pods cannot share the same port on the same node.

## How does it work?

### MinReadySeconds for StatefulSets

The StatefulSet controller watches for the StatefulSet pods and counts how long a particular pod has been in the `Running` state, if this value is greater than or equal to the time specified in `.spec.minReadySeconds` field of the StatefulSet, the StatefulSet controller updates the `AvailableReplicas` field in the StatefulSet's status.

### MaxSurge for DaemonSets

> The DaemonSet controller creates the additional pods (above the desired number resulting from DaemonSet spec) based on the value given in `.spec.strategy.rollingUpdate.maxSurge`. The additional pods would run on the same node where the old DaemonSet pod is running till the old pod gets killed. This value cannot be `0` when `MaxUnavailable` is 0. The default value is 0.

Reviewer: It would be nice to outline the fact that `maxSurge` can be specified either as a absolute number of pods, or a percentage of desired pods.

## How do I use it?

### MinReadySeconds for StatefulSets

Specify a value for `minReadySeconds` for any StatefulSet and check if pods are available or not by inspecting
`AvailableReplicas` field using:

`kubectl get statefulset/<name_of_the_statefulset> -o yaml`

Please note that the default value of `minReadySeconds` is 0.

### MaxSurge for DaemonSets

Specify a value for `.spec.updateStrategy.rollingUpdate.maxSurge` and set `.spec.updateStrategy.rollingUpdate.maxUnavailable` to `0`. 

Then observe a faster rollout and higher number of pods running at the same time in the next rollout.

```

> oc rollout restart daemonset <name_of_the_daemonset>

Reviewer 2: Is this an openshift-specific feature? If not, then could you show this with a kubectl command instead of an oc command?

Author: It is not openshift specific. I forgot to switch it back to kubectl
